pkg/certgenerator/v1beta1/generator.go

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2022 The Kubeflow Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certgenerator
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"errors"
+	"fmt"
+	"math/big"
+	"strings"
+	"time"
+
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	configv1beta1 "github.com/kubeflow/katib/pkg/apis/config/v1beta1"
+	"github.com/kubeflow/katib/pkg/controller.v1beta1/consts"
+)
+
+var (
+	errServiceNotFound      = errors.New("unable to locate controller service")
+	errCertCheckFail        = errors.New("failed to check if certs already exist")
+	errCreateCertFail       = errors.New("failed to create certs")
+	errCreateCertSecretFail = errors.New("failed to create secret embedded certs")
+	errInjectCertError      = errors.New("failed to inject certs into WebhookConfigurations")
+)
+
+// CertGenerator is the manager to generate certs.
+type CertGenerator struct {
+	namespace   string
+	serviceName string
+	secretName  string
+	kubeClient  client.Client
+	certsReady  chan struct{}
+
+	certs             *certificates
+	fullServiceDomain string
+}
+
+var _ manager.Runnable = &CertGenerator{}
+var _ manager.LeaderElectionRunnable = &CertGenerator{}
+
+func (c *CertGenerator) Start(ctx context.Context) error {
+	if err := c.generate(ctx); err != nil {
+		return err
+	}
+	// Close a certsReady means start to register controllers to the manager.
+	close(c.certsReady)
+	return nil
+}
+
+func (c *CertGenerator) NeedLeaderElection() bool {
+	return true
+}
```

Comment:
Do we need this ?

Reply:
Yes, we need this. By implementing the `LeaderElectionRunnable` interface, the manager applies leader election logic to the cert-generator.

https://github.com/kubeflow/katib/blob/eac65bf8e31fe4d3a2c4dbb20be30d600c086157/pkg/certgenerator/v1beta1/generator.go#L63

Reply:
@tenzen-y Why we didn't add those [runnable to our Experiment, Trial, and Suggestion controllers?](https://github.com/kubeflow/katib/blob/master/pkg/controller.v1beta1/experiment/experiment_controller.go)

Reply:
@andreyvelich The controllers, such as the experiment-controller implements the `Reconciler` interface, not the `Runnable` interface.

https://github.com/kubeflow/katib/blob/74cf5b8d4ecc3666db3966a99b66ed17e8c69129/pkg/controller.v1beta1/experiment/experiment_controller.go#L136

Reply:
Oh, yeah, you are right. So we need to configure leader election only for Runnable.

Reply:
Correct.